File: src/main/java/com/dude/dms/backend/data/rules/Rule.java
package com.dude.dms.backend.data.rules;

import com.dude.dms.backend.data.DataEntity;
import com.dude.dms.backend.data.Tag;

import javax.persistence.MappedSuperclass;
import java.util.Set;

@MappedSuperclass
public abstract class Rule extends DataEntity {

    protected Boolean active;

    /**
     * Checks a line of text against the rule pattern.
     * @param line line of text
     * @return {@code true}, if the pattern found a match
     */
    public abstract boolean validate(String line);

    public abstract Set<Tag> getTags();

    public abstract void setTags(Set<Tag> tags);

    protected Rule() {
    }

    protected Rule(Boolean active) {
        this.active = active;
    }

    public Boolean getActive() {
        return active;
    }

    public void setActive(Boolean active) {
        this.active = active;
    }
}

File: src/main/java/com/dude/dms/backend/data/rules/Rule.kt
package com.dude.dms.backend.data.rules

import com.dude.dms.backend.data.DataEntity
import com.dude.dms.backend.data.LogsEvents
import com.dude.dms.backend.data.Tag
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class Rule : DataEntity(), LogsEvents {

    abstract var tags: Set<Tag>

    override fun showEvents() = true

    /**
     * Checks a line of text against the rule pattern.
     * @param line line of text
     * @return `true`, if the pattern found a match
     */
    abstract fun validate(line: String?): Boolean
}